polkadot/node/core/candidate-validation/src/lib.rs

```diff
@@ -427,14 +427,15 @@ where
 		.iter()
 		.any(|v| keystore.has_keys(&[(v.to_raw_vec(), AuthorityDiscoveryId::ID)]));
 
-	let is_present_authority = session_info
-		.discovery_keys
+	// We could've checked discovery_keys but on Kusama validators.len() < discovery_keys.len().
+	let is_present_validator = session_info
+		.validators
 		.iter()
-		.any(|v| keystore.has_keys(&[(v.to_raw_vec(), AuthorityDiscoveryId::ID)]));
+		.any(|v| keystore.has_keys(&[(v.to_raw_vec(), ValidatorId::ID)]));
 
 	// There is still a chance to be a previous session authority, but this extra work does not
 	// affect the finalization.
-	is_past_present_or_future_authority && !is_present_authority
+	is_past_present_or_future_authority && !is_present_validator
```

Comment:
This distinction exists until we remove the `max_validators` configuration so even without this fix it still works as expected on Polkadot.

Reply:
Sure